src/gui/src/scriptWidget.cpp

```diff
@@ -401,30 +408,39 @@ class ScriptWidget::GuiSink : public spdlog::sinks::base_sink<Mutex>
   {
     // Convert the msg into a formatted string
     spdlog::memory_buf_t formatted;
+
+    mutex_.lock(); // formatter checks and caches some information
     this->formatter_->format(msg, formatted);
-    std::string formatted_msg = std::string(formatted.data(), formatted.size());
+    mutex_.unlock();
+    const QString formatted_msg = QString::fromStdString(std::string(formatted.data(), formatted.size()));
 
     if (msg.level == spdlog::level::level_enum::off) {
       // this comes from a ->report
-      widget_->addReportToOutput(formatted_msg.c_str());
+      widget_->addReportToOutput(formatted_msg);
     }
     else {
       // select error message color if message level is error or above.
       const QColor& msg_color = msg.level >= spdlog::level::level_enum::err ? widget_->tcl_error_msg_ : widget_->buffer_msg_;
 
-      widget_->addLogToOutput(formatted_msg.c_str(), msg_color);
+      widget_->addLogToOutput(formatted_msg, msg_color);
     }
+
+    // process queue, if main thread will process new text, otherwise there is nothing to process from this thread.
+    QCoreApplication::processEvents(QEventLoop::ExcludeUserInputEvents);
   }
 
   void flush_() override {}
 
  private:
   ScriptWidget* widget_;
+  std::mutex mutex_;
 };
 
 void ScriptWidget::setLogger(utl::Logger* logger)
 {
-  sink_ = std::make_shared<GuiSink<std::mutex>>(this);
+  // use spdlog::details::null_mutex instead of std::mutex, Qt will handle the thread transfers to the output viewer
+  // null_mutex prevents deadlock (by not locking) when the logging causes a redraw, which then causes another logging event.
```

Comment:
Is it the call to addReportToOutput/addLogToOutput that triggers the new output?  If so, are we calling the widget while in the middle of another call?

Reply:
Not sure what you mean? `addTextToOutput` this function is the one that adds new output the output widget (when adding the new text it triggers another paint event).

The calls to `addReportToOutput`/`addLogToOutput` (each end up calling `addToOutput`). This change makes `addToOutput` a signal, which is connected via a `QueuedConnection` to `addTextToOutput`. Using the queued connected allows Qt to handle the moving the data between threads and ensures only the main GUI thread is interacting with the Qt objects.

Reply:
I forgot to discuss this at our meeting.  I'm not sure I get the queued connection part of your comment - could you elaborate or should we do a quick  call?